test/cppsuite/test_harness/connection.h

```diff
@@ -0,0 +1,109 @@
+/*-
+ * Public Domain 2014-present MongoDB, Inc.
+ * Public Domain 2008-2014 WiredTiger, Inc.
+ *
+ * This is free and unencumbered software released into the public domain.
+ *
+ * Anyone is free to copy, modify, publish, use, compile, sell, or
+ * distribute this software, either in source code form or as a compiled
+ * binary, for any purpose, commercial or non-commercial, and by any
+ * means.
+ *
+ * In jurisdictions that recognize copyright laws, the author or authors
+ * of this software dedicate any and all copyright interest in the
+ * software to the public domain. We make this dedication for the benefit
+ * of the public at large and to the detriment of our heirs and
+ * successors. We intend this dedication to be an overt act of
+ * relinquishment in perpetuity of all present and future rights to this
+ * software under copyright law.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#ifndef CONN_API_H
+#define CONN_API_H
+
+#include <mutex>
+
+extern "C" {
+#include "test_util.h"
+#include "wiredtiger.h"
+}
+
+#include "api_const.h"
+#include "debug_utils.h"
+
+namespace test_harness {
```

Comment:
I'm a bit concerned we define connection as a singleton. I think it is a valid use case to operate on two different databases. e.g., one is a copy of the original database. We do this a lot in the python test already.

Reply:
I think that's a valid concern, so some possible solutions are:
- An interface connection_manager which is implemented by the test and gives it a single connection instances, then the connection manager can be passed to components that may need a connection.
- A connection wrapper which is just passed between objects

If we do end up having multiple connections then I think we'll have other issues with regards to knowing which URI is for which connection etc. The main reason for the singleton was to avoid concurrency issues with accessing connection API calls.

Reply:
I don't understand how singleton solves the concurrency issue. I see we are using a mutex in session to control concurrency. How is the singleton related then?

Reply:
Well actually yeah you're correct, the solution to the concurrency issue would be the same even if we passed a reference to a class that held a connection. It would still have a single mutex. The singleton was a lazy way of avoid dependency injection. What would you prefer?

Reply:
I think we shouldn't make it a singleton and we should just pass its reference around to whichever class needs it.

Reply:
Okay well Etienne and I would perter to keep the singleton for now and if the requirements change we'll either pass Connection container object or similar.

Reply:
A singleton could manage different types of connection. Passing the reference around seems to be a lot as it is used in quite a few places already and might be used even more in the next dev.

Reply:
Thinking longer term however if we were to manage multiple connections would we have an array of Connection wrappers? I suppose so, the index into the array could be the usecase for the connection.

Reply:
Yes that's correct, a map/an array, any suitable container to store multiple connections. An ID could be returned upon creation of pre defined IDs as we discussed.